### PR TITLE
Add UIE

### PR DESCRIPTION
**Pull request by linjieccc**
### PR types

New features
### PR changes

APIs
### Description

Add UIE

### CODE REVIEW

File: examples/information_extraction/uie/README.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# UIE
+
+Universal Information Extraction（UIE）将各种类型的信息抽取任务统一转化为自然语言的形式，并进行多任务联合训练。该模型支持多种类型的开放抽取任务，包括但不限于命名实体、关系、事件论元、事件描述片段、评价、评价维度、观点词、情感倾向等。
```
Comment (ZeyuChen) [inline examples/information_extraction/uie/README.md]: 最好README简洁得加上对应的Paper链接介绍

File: examples/information_extraction/uie/run_evaluate.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+python evaluate.py \
+    --model_path "./checkpoint/model_best/model_state.pdparams" \
+    --test_path "./data/ext_data/test.txt" \
+    --batch_size 16 \
+    --max_seq_len 512
```
Comment (wawltor) [inline examples/information_extraction/uie/run_evaluate.sh]: 去掉shell脚本

File: examples/information_extraction/uie/run_train.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    --save_dir "./checkpoint" \
+    --learning_rate 1e-5 \
+    --batch_size 16 \
+    --max_seq_len 512 \
+    --num_epochs 50 \
+    --init_from_ckpt "./uie_model/model_state.pdparams" \
+    --seed 1000 \
+    --logging_steps 10 \
+    --valid_steps 100 \
+    --device "gpu"
```
Comment (wawltor) [inline examples/information_extraction/uie/run_train.sh]: 去掉shell脚本

File: examples/information_extraction/uie/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+通过`schema`自定义抽取目标，`task_path`指定使用标注数据训练的UIE模型。
+
+```python
+from paddlenlp import Taskflow
+
+schema = [{"作品名": ["作者", "出版社名称"]}]
+
+# 为任务实例设定抽取目标和定制化模型权重路径
+my_ie = Taskflow("information_extraction", schema=schema, task_path='./checkpoint/model_best')
+```
```
Comment (wawltor) [inline examples/information_extraction/uie/README.md:210]: 文档这块非常重要，需要写一个细致的文档大纲，然后进行讨论

File: examples/information_extraction/uie/model.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            token_type_ids=token_type_ids,
+            position_ids=pos_ids,
+            attention_mask=att_mask)
+        start_logits = self.linear_start(sequence_output)
+        start_logits = paddle.squeeze(start_logits, -1)
+        start_prob = self.sigmoid(start_logits)
+        end_logits = self.linear_end(sequence_output)
+        end_logits = paddle.squeeze(end_logits, -1)
+        end_prob = self.sigmoid(end_logits)
+        return start_prob, end_prob
```
Comment (wawltor) [inline examples/information_extraction/uie/model.py:52]: 这里有个疑问，是不是可以直接去加载finetune后的模型，因为这里本就是few-shot，baseline模型应该一个训练后的模型

File: examples/information_extraction/uie/train.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    tokenizer = AutoTokenizer.from_pretrained("ernie-1.0")
+    model = UIE()
+
+    if args.init_from_ckpt and os.path.isfile(args.init_from_ckpt):
+        state_dict = paddle.load(args.init_from_ckpt)
+        model.set_dict(state_dict)
+        print("Init from: {}".format(args.init_from_ckpt))
+
+    train_ds = load_dataset(reader, data_path=args.train_path, lazy=False)
+    dev_ds = load_dataset(reader, data_path=args.dev_path, lazy=False)
```
Comment (wawltor) [inline examples/information_extraction/uie/train.py]: 这块的数据切分数据是什么策略了

File: examples/information_extraction/uie/train.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+            optimizer.clear_grad()
+            loss_list.append(float(loss))
+
+            global_step += 1
+            if global_step % args.logging_steps == 0 and rank == 0:
+                time_diff = time.time() - tic_train
+                loss_avg = sum(loss_list) / len(loss_list)
+                num_correct, num_infer, num_label = get_metric(
+                    start_prob, end_prob, start_ids, end_ids)
+                precision, recall, f1 = get_f1(num_correct, num_infer,
```
Comment (wawltor) [inline examples/information_extraction/uie/train.py]: f1的计算，PaddleNLP的内部是有相应的metric的

File: examples/information_extraction/uie/train.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    parser.add_argument("--batch_size", default=16, type=int, help="Batch size per GPU/CPU for training.")
+    parser.add_argument("--learning_rate", default=1e-5, type=float, help="The initial learning rate for Adam.")
+    parser.add_argument("--train_path", default=None, type=str, help="The path of train set.")
+    parser.add_argument("--dev_path", default=None, type=str, help="The path of dev set.")
+    parser.add_argument("--save_dir", default='./checkpoint', type=str, help="The output directory where the model checkpoints will be written.")
+    parser.add_argument("--max_seq_len", default=512, type=int, help="The maximum total input sequence length after tokenization. "
+        "Sequences longer than this will be truncated, sequences shorter will be padded.")
+    parser.add_argument("--num_epochs", default=50, type=int, help="Total number of training epochs to perform.")
+    parser.add_argument("--init_from_ckpt", default="uie/model_state.pdparams", type=str, help="The path of checkpoint to be loaded.")
+    parser.add_argument("--seed", default=1000, type=int, help="random seed for initialization")
```
Comment (wawltor) [inline examples/information_extraction/uie/train.py]: random -> Random

File: examples/information_extraction/uie/evaluate.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        num_correct += res[0]
+        num_infer += res[1]
+        num_label += res[2]
+    precision, recall, f1 = get_f1(num_correct, num_infer, num_label)
+    model.train()
+    return precision, recall, f1
+
+
+def do_eval():
+    tokenizer = AutoTokenizer.from_pretrained("ernie-1.0")
```
Comment (wawltor) [inline examples/information_extraction/uie/evaluate.py]: 这块的模型名字不要限制，后续还要切换模型，有ERNIE-3.0一系列模型，还有垂类模型

File: examples/information_extraction/uie/train.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+def do_train():
+    paddle.set_device(args.device)
+    rank = paddle.distributed.get_rank()
+    if paddle.distributed.get_world_size() > 1:
+        paddle.distributed.init_parallel_env()
+
+    set_seed(args.seed)
+
+    tokenizer = AutoTokenizer.from_pretrained("ernie-1.0")
```
Comment (wawltor) [inline examples/information_extraction/uie/train.py]: 这块的模型名字要提供可选项

File: examples/information_extraction/uie/utils.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wawltor) [inline examples/information_extraction/uie/utils.py]: 2020->2022

File: examples/information_extraction/uie/utils.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+import json
+
+import paddle
```
Comment (wawltor) [inline examples/information_extraction/uie/utils.py]: paddle 往下摞

File: examples/information_extraction/uie/doccano.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import argparse
+
```
Comment (wawltor) [inline examples/information_extraction/uie/doccano.py]: 不要空行

File: examples/information_extraction/uie/doccano.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+from utils import set_seed, save_examples, convert_doccano_examples
+
+
+def doccano2SA(doccano_file,
+               save_dir,
+               splits=[0.8, 0.9],
+               negative_ratio=1,
+               is_shuffle=True):
+    """
+        @Description: Consvert doccano file to data format which is suitable to input to this Application.
```
Comment (wawltor) [inline examples/information_extraction/uie/doccano.py]: 这块的参数注释，注意一下规范

File: examples/information_extraction/uie/doccano.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+import os
+import argparse
+
+import numpy as np
+
+from utils import set_seed, save_examples, convert_doccano_examples
+
+
+def doccano2SA(doccano_file,
```
Comment (wawltor) [inline examples/information_extraction/uie/doccano.py]: SA是啥意思？

File: examples/information_extraction/uie/metric.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+    pred_start_ids = get_bool_ids_greater_than(start_prob)
+    pred_end_ids = get_bool_ids_greater_than(end_prob)
+    gold_start_ids = get_bool_ids_greater_than(start_ids.tolist())
+    gold_end_ids = get_bool_ids_greater_than(end_ids.tolist())
+    num_correct = 0
+    num_infer = 0
+    num_label = 0
+    for predict_start_ids, predict_end_ids, label_start_ids, label_end_ids in zip(
+            pred_start_ids, pred_end_ids, gold_start_ids, gold_end_ids):
+        [_correct, _infer, _label] = eval_span(
```
Comment (wawltor) [inline examples/information_extraction/uie/metric.py]: 这块的eval_span再看看，metric中理论上train eval都可以适用的

File: examples/information_extraction/uie/doccano.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+    args = parser.parse_args()
+    # yapf: enable
+
+    # Ensure generate the same negative samples for one seed.
+    set_seed(1000)
+
+    doccano2SA(
+        args.doccano_file,
+        args.save_dir,
+        splits=[0.6, 0.8],
```
Comment (wawltor) [inline examples/information_extraction/uie/doccano.py]: 0.6 + 0.8  > 1.0 有点奇怪

File: examples/information_extraction/uie/utils.py [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+
+def save_examples(examples, save_path, idxs):
+    with open(save_path, "w", encoding="utf-8") as f:
+        for idx in idxs:
+            for example in examples[idx]:
+                line = json.dumps(example, ensure_ascii=False) + "\n"
+                f.write(line)
+
+
+def add_negtive_example(examples, texts, prompts, label_set, negative_ratio):
```
Comment (wawltor) [inline examples/information_extraction/uie/utils.py]: 这里有个疑问，因为这里做了一个笛卡尔积，怎么保证负样本不在正样本集合里面了？

File: examples/information_extraction/uie/utils.py [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+                    result)
+
+            if predicate not in predicate_set:
+                predicate_set.append(predicate)
+            relation_prompt.append(relation_label)
+
+        for v in relation_example_map.values():
+            relation_example.append(v)
+
+        relation_examples.append(relation_example)
```
Comment (wawltor) [inline examples/information_extraction/uie/utils.py]: 分类样本怎么构建出来的了？

File: examples/information_extraction/uie/README.md (hunk 1 above)
Comment (linjieccc) [reply examples/information_extraction/uie/README.md]: 这里UIE中文版使用的理解式模型，与paper的生成式版本有所区别，这块我准备个文档大纲，然后线下讨论下

**Review by wawltor — DISMISSED** [review]
LGTM

File: examples/information_extraction/uie/model.py (hunk 5 above)
Comment (linjieccc) [reply examples/information_extraction/uie/model.py:52]: done，已修改为直接下载训练后的模型的形式

File: examples/information_extraction/uie/train.py (hunk 6 above)
Comment (linjieccc) [reply examples/information_extraction/uie/train.py]: 目前这里是按max_seq_len进行切分，目的是支持用户在doccano数据标注阶段直接以文档形式进行标注

File: examples/information_extraction/uie/train.py (hunk 8 above)
Comment (linjieccc) [reply examples/information_extraction/uie/train.py]: done

File: examples/information_extraction/uie/doccano.py (hunk 15 above)
Comment (linjieccc) [reply examples/information_extraction/uie/doccano.py]: 已修改函数名为convert_doccano_file

File: examples/information_extraction/uie/utils.py (hunk 18 above)
Comment (linjieccc) [reply examples/information_extraction/uie/utils.py]: 这里会遍历每一条数据（对应有一个独立的prompt_set），每一次都对独立的prompt_set和所有的label_set取一次差集

**Review by wawltor — DISMISSED** [review]
LGTM

File: examples/information_extraction/uie/README.md [hunk 20]
```diff
@@ -0,0 +1,182 @@
+# 通用信息抽取UIE
```
Comment (ZeyuChen) [inline examples/information_extraction/uie/README.md]: UIE缩写最好也解释下吧
通用信息抽取 Universal Information Extraction (UIE)

File: examples/information_extraction/uie/README.md [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+* [2. 模型简介](#模型简介)
+* [3. 应用场景](#应用场景)
+* [4. 开箱即用](#开箱即用)
+* [5. 轻定制功能](#轻定制功能)
+
+<a name="模型简介"></a>
+
+## 1. 功能简介
+
+开放域信息抽取是信息抽取的一种全新范式，主要思想是减少人工参与，利用单一模型支持多种类型的开放抽取任务，用户可以使用自然语言自定义抽取目标，在实体、关系类别等未定义的情况下抽取输入文本中的信息片段。
```
Comment (ZeyuChen) [inline examples/information_extraction/uie/README.md]: 建议这段介绍，摘取原论文的背景介绍进行翻译。这段话显得比较苍白无力。

File: paddlenlp/utils/tools.py [hunk 22]
```diff
@@ -162,3 +162,79 @@ def compare_version(version, pair_version):
         elif int(version_code) < int(pair_version_code):
             return -1
     return 0
+
+
+def get_bool_ids_greater_than(probs, limit=0.5, return_prob=False):
+    """
+    get idx of the last dim in prob arraies, which is greater than a limitation
```
Comment (ZeyuChen) [inline paddlenlp/utils/tools.py]: 所有的注释一律按大写字母开头标准的英文语法注释。然后检查这里的英文语法吧。arraies是typo

File: examples/information_extraction/uie/utils.py [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+            entity_prompts.append(entity_prompt)
+
+            relation_example = []
+            relation_prompt = []
+            relation_example_map = {}
+            for relation in relations:
+                predicate = relation["type"]
+                subject_id = relation["from_id"]
+                object_id = relation["to_id"]
+                relation_label = entity_map[subject_id][
```
Comment (ZeyuChen) [inline examples/information_extraction/uie/utils.py]: 这么tricky的代码，是否可以注释下？

**Review by ZeyuChen — COMMENTED** [review]
完善下代码注释，统一下英文注释语法风格。
README逻辑和介绍完善。

File: examples/information_extraction/uie/train.py [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+    model = UIE(encoding_model, hidden_size)
+
+    pretrained_model_path = os.path.join(args.model, "model_state.pdparams")
+
+    if os.path.exists(pretrained_model_path):
+        state_dict = paddle.load(pretrained_model_path)
+        model.set_dict(state_dict)
+        print("Init from: {}".format(pretrained_model_path))
+    else:
+        get_path_from_url(url, args.model)
```
Comment (LemonNoel) [inline examples/information_extraction/uie/train.py]: 这里当初始化模型参数不存在时，仅会下载参数而没有真的初始化，导致模型会train from scratch。

Comment (LemonNoel) [inline examples/information_extraction/uie/train.py:53]: 这里当初始化模型参数不存在时，仅会下载参数而没有真的初始化，导致模型会train from scratch。

File: examples/information_extraction/uie/utils.py (hunk 19 above)
Comment (linjieccc) [reply examples/information_extraction/uie/utils.py]: 已补充分类任务

File: examples/information_extraction/uie/README.md (hunk 20 above)
Comment (linjieccc) [reply examples/information_extraction/uie/README.md]: done, thx

File: examples/information_extraction/uie/README.md (hunk 21 above)
Comment (linjieccc) [reply examples/information_extraction/uie/README.md]: done

File: examples/information_extraction/uie/utils.py (hunk 23 above)
Comment (linjieccc) [reply examples/information_extraction/uie/utils.py]: 已补充注释说明

File: paddlenlp/utils/tools.py (hunk 22 above)
Comment (linjieccc) [reply paddlenlp/utils/tools.py]: done, thx

File: examples/information_extraction/uie/train.py (hunk 24 above)
Comment (linjieccc) [reply examples/information_extraction/uie/train.py]: done, thx

File: examples/information_extraction/uie/doccano.md [hunk 25]
```diff
@@ -0,0 +1,278 @@
...
+## 3. 数据上传
+
+上传的文件为txt格式，每一行为一条待标注文本，示例：
+
+```text
+2月8日上午北京冬奥会自由式滑雪女子大跳台决赛中中国选手谷爱凌以188.25分获得金牌
+第十四届全运会在西安举办
+```
+
+上传数据类型选择``TextLine``
```
Comment (wawltor) [inline examples/information_extraction/uie/doccano.md]: 这里具体再说明一下TextLine和TextFile，看起来踩坑的人特别多
重点加粗

Comment (linjieccc) [reply examples/information_extraction/uie/doccano.md]: done

**Review by wawltor — APPROVED** [review]
LGTM
